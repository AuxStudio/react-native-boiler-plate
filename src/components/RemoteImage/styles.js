import { StyleSheet } from 'react-native';

import styleConstants from '../../styleConstants';

const styles = StyleSheet.create({
  container: {},
  image: {},
  backgroundContainer: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: styleConstants.colors.dividerColor,
  },
  icon: {
    fontSize: styleConstants.fonts.sizes.icon,
    color: styleConstants.colors.primaryText,
  },
});

export default styles;
