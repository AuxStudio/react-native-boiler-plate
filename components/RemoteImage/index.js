import React from 'react';
import PropTypes from 'prop-types';
import { View, ViewPropTypes, Text, ActivityIndicator } from 'react-native';
import FastImage from 'react-native-fast-image';
import Icon from 'react-native-vector-icons/MaterialIcons';

import styles from './styles';
import styleConstants from '../../styleConstants';

export default class RemoteImage extends React.Component {
  constructor(props) {
    super(props);

    this.setError = this.setError.bind(this);
    this.setLoading = this.setLoading.bind(this);

    this.state = {
      isLoading: true,
      hasError: false,
    };
  }

  static propTypes = {
    source: PropTypes.shape({
      uri: PropTypes.string,
    }),
    borderRadius: PropTypes.number,
    iconStyle: Text.propTypes.style,
    style: ViewPropTypes.style,
    loaderColor: PropTypes.string,
  };

  static defaultProps = {
    loaderColor: styleConstants.colors.primaryText,
  };

  setError() {
    this.setState({
      hasError: true,
    });

    this.setLoading(false);
  }

  setLoading(isLoading) {
    this.setState({
      isLoading,
    });
  }

  render() {
    const borderRadiusStyles = this.props.borderRadius && { borderRadius: this.props.borderRadius };
    let backgroundComponent;

    if (this.state.hasError) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <Icon name="error-outline" style={[styles.icon, this.props.iconStyle]} />
        </View>
      );
    } else if (this.state.isLoading) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <ActivityIndicator size="large" color={this.props.loaderColor} />
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <FastImage
          source={this.props.source}
          style={[styles.image, borderRadiusStyles, this.props.style]}
          onLoadEnd={() => this.setLoading(false)}
          onError={this.setError}
        />
        {backgroundComponent}
      </View>
    );
  }
}
