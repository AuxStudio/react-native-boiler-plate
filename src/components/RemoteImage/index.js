import React from 'react';
import PropTypes from 'prop-types';
import { View, ViewPropTypes, Text, ActivityIndicator } from 'react-native';
import FastImage from 'react-native-fast-image';
import Icon from 'react-native-vector-icons/MaterialIcons';

import styles from './styles';
import styleConstants from '../../styleConstants';

export default class RemoteImage extends React.Component {
  constructor(props) {
    super(props);

    this.setHasError = this.setHasError.bind(this);
    this.setIsLoading = this.setIsLoading.bind(this);

    this.state = {
      isLoading: true,
      hasError: false,
    };
  }

  static propTypes = {
    source: PropTypes.shape({
      uri: PropTypes.string,
    }),
    borderRadius: PropTypes.number,
    iconStyle: Text.propTypes.style,
    style: ViewPropTypes.style,
    loaderColor: PropTypes.string,
  };

  static defaultProps = {
    loaderColor: styleConstants.colors.primaryText,
  };

  setHasError() {
    this.setState({
      hasError: true,
    });

    this.setIsLoading(false);
  }

  setIsLoading(isLoading) {
    this.setState({
      isLoading,
    });
  }

  render() {
    const { hasError, isLoading } = this.state;
    const { borderRadius, iconStyle, loaderColor, source, style } = this.props;
    const borderRadiusStyles = borderRadius && { borderRadius };
    let backgroundComponent;

    if (hasError) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <Icon name="error-outline" style={[styles.icon, iconStyle]} />
        </View>
      );
    } else if (isLoading) {
      backgroundComponent = (
        <View style={[styles.backgroundContainer, borderRadiusStyles]}>
          <ActivityIndicator size="large" color={loaderColor} />
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <FastImage
          source={source}
          style={[styles.image, borderRadiusStyles, style]}
          onLoadEnd={() => this.setIsLoading(false)}
          onError={this.setHasError}
        />
        {backgroundComponent}
      </View>
    );
  }
}
