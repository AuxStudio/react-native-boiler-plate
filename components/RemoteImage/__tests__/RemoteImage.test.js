import React from 'react';
import renderer from 'react-test-renderer';

import RemoteImage from '..';

const IMAGE_SOURCE = 'https://d3iw72m71ie81c.cloudfront.net/male-47.jpg';

describe('RemoteImage', () => {
  it('renders with all props', () => {
    expect(
      renderer.create(
        <RemoteImage
          source={{ uri: IMAGE_SOURCE }}
          borderRadius={10}
          iconStyle={{ color: 'red' }}
          style={{ backgroundColor: 'blue' }}
          loaderColor="green"
        />,
      ),
    ).toMatchSnapshot();
  });

  it('renders with minimum required props', () => {
    expect(renderer.create(<RemoteImage />)).toMatchSnapshot();
  });

  it('renders and toggles loading', () => {
    const component = renderer.create(<RemoteImage source={{ uri: IMAGE_SOURCE }} />);
    const instance = component.getInstance();

    instance.setLoading(false);

    expect(instance.state.isLoading).toBe(false);
    expect(component).toMatchSnapshot();
  });

  it('renders with error', () => {
    const component = renderer.create(<RemoteImage source={{ uri: IMAGE_SOURCE }} />);
    const instance = component.getInstance();

    instance.setError();

    expect(instance.state.hasError).toBe(true);
    expect(instance.state.isLoading).toBe(false);
    expect(component).toMatchSnapshot();
  });
});
