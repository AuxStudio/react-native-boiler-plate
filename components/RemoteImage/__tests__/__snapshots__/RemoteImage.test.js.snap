// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoteImage renders and toggles loading 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Array [
        Array [
          Object {},
          undefined,
          undefined,
        ],
        Object {
          "overflow": "hidden",
        },
      ]
    }
  >
    <FastImageView
      onFastImageError={[Function]}
      onFastImageLoadEnd={[Function]}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://d3iw72m71ie81c.cloudfront.net/male-47.jpg",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
</View>
`;

exports[`RemoteImage renders with all props 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Array [
        Array [
          Object {},
          Object {
            "borderRadius": 10,
          },
          Object {
            "backgroundColor": "blue",
          },
        ],
        Object {
          "overflow": "hidden",
        },
      ]
    }
  >
    <FastImageView
      onFastImageError={[Function]}
      onFastImageLoadEnd={[Function]}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://d3iw72m71ie81c.cloudfront.net/male-47.jpg",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0E0E0",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "borderRadius": 10,
        },
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="green"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`RemoteImage renders with error 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Array [
        Array [
          Object {},
          undefined,
          undefined,
        ],
        Object {
          "overflow": "hidden",
        },
      ]
    }
  >
    <FastImageView
      onFastImageError={[Function]}
      onFastImageLoadEnd={[Function]}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://d3iw72m71ie81c.cloudfront.net/male-47.jpg",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0E0E0",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Array [
            Object {
              "color": "#212121",
              "fontSize": 24,
            },
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      ÓÄÅ
    </Text>
  </View>
</View>
`;

exports[`RemoteImage renders with minimum required props 1`] = `
<View
  style={Object {}}
>
  <Image
    onError={[Function]}
    onLoadEnd={[Function]}
    resizeMode="cover"
    style={
      Array [
        Object {},
        undefined,
        undefined,
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0E0E0",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#212121"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;
